x-pack/plugin/versionfield/src/main/java/org/elasticsearch/xpack/versionfield/VersionStringFieldMapper.java

```diff
@@ -0,0 +1,419 @@
+        }
+    }
+
+    static class Builder extends ParametrizedFieldMapper.Builder {
+
+        private final Parameter<Map<String, String>> meta = Parameter.metaParam();
+
+        Builder(String name) {
+            super(name);
+        }
+
+        private VersionStringFieldType buildFieldType(BuilderContext context, FieldType fieldtype) {
+            return new VersionStringFieldType(buildFullName(context), fieldtype, meta.getValue());
+        }
+
+        @Override
+        public VersionStringFieldMapper build(BuilderContext context) {
+            FieldType fieldtype = new FieldType(Defaults.FIELD_TYPE);
+            return new VersionStringFieldMapper(
+                name,
+                fieldtype,
+                buildFieldType(context, fieldtype),
+                multiFieldsBuilder.build(this, context),
+                copyTo.build()
+            );
+        }
+
+        @Override
+        protected List<Parameter<?>> getParameters() {
+            return List.of(meta);
+        }
+    }
+
+    public static final TypeParser PARSER = new TypeParser((n, c) -> new Builder(n));
+
+    public static final class VersionStringFieldType extends TermBasedFieldType {
+
+        public VersionStringFieldType(String name, FieldType fieldType, Map<String, String> meta) {
+            super(name, true, true, new TextSearchInfo(fieldType, null, Lucene.KEYWORD_ANALYZER, Lucene.KEYWORD_ANALYZER), meta);
+            setIndexAnalyzer(Lucene.KEYWORD_ANALYZER);
+        }
+
+        @Override
+        public String typeName() {
+            return CONTENT_TYPE;
+        }
+
+        @Override
+        public Query existsQuery(QueryShardContext context) {
+            return new DocValuesFieldExistsQuery(name());
+        }
+
+        @Override
+        public Query prefixQuery(String value, MultiTermQuery.RewriteMethod method, QueryShardContext context) {
+            if (context.allowExpensiveQueries() == false) {
+                throw new ElasticsearchException(
+                    "[prefix] queries cannot be executed when '"
+                        + ALLOW_EXPENSIVE_QUERIES.getKey()
+                        + "' is set to false. For optimised prefix queries on text "
+                        + "fields please enable [index_prefixes]."
+                );
+            }
+            return wildcardQuery(value + "*", method, context);
+        }
+
+        /**
+         * We cannot simply use RegexpQuery directly since we use the encoded terms from the dictionary, but the
+         * automaton in the query will assume unencoded terms. We are running through all terms, decode them and
+         * then run them through the automaton manually instead. This is not as efficient as intersecting the original
+         * Terms with the compiled automaton, but we expect the number of distinct version terms indexed into this field
+         * to be low enough and the use of "regexp" queries on this field rare enough to brute-force this
```

Comment:
May be we can add to do for subsequent PRs to optimize these queries

Reply:
I'm not a bit fan of tracking backlog like this in TODOs in code, I'd rather open an issue to investigate improvements when time allows and track it with any other issue we have. Would that be okay for you?

Reply:
@cbuescher Thanks, that would be fine.

Reply:
I opened https://github.com/elastic/elasticsearch/issues/62608 for tracking